rfd/00XX-agent-census.md

```diff
@@ -0,0 +1,237 @@
+8. Container runtime (e.g. Docker)
+9. Container orchestrator (e.g. Kubernetes)
+10. Cloud environment (e.g. AWS, GCP, Azure)
+
+#### Data collection
+
+Currently, when an agent first starts, the [inventory control system](https://github.com/gravitational/teleport/tree/6f9ad9553a5b5946f57cb35411c598754d3f926b/lib/inventory) sends an [`UpstreamInventoryHello`](https://github.com/gravitational/teleport/blob/6f9ad9553a5b5946f57cb35411c598754d3f926b/api/proto/teleport/legacy/client/proto/authservice.proto#L1936-L1953) message to the auth server.
+This message has the following fields:
+
+```protobuf
+message UpstreamInventoryHello {
+  string Version = 1;
+  string ServerID = 2;
+  repeated string Services = 3 [(gogoproto.casttype) = "github.com/gravitational/teleport/api/types.SystemRole"];
+  string Hostname = 4;
+}
+```
+
+The `Version` field contains the Teleport version, while the `Services` field contains the subset of the system roles that are currently active at the agent.
+
+We will extend this message to contain the remaining agent data that we want to track:
+
+```protobuf
+message UpstreamInventoryHello {
+  // (...)
+  string OS = 5;
+  string OSVersion = 6;
+  string HostArchitecture = 7;
+  string GLibCVersion = 8;
+  repeated string InstallMethods = 9;
+  string ContainerRuntime = 10;
+  string ContainerOrchestrator = 11;
+  string CloudEnvironment = 12;
+}
+```
+
+When the auth server receives an `UpstreamInventoryHello` message, it will take the information in the message and send it to PreHog.
+For this, a new PreHog `AgentMetadataEvent` message will be added (note that only the `UpstreamInventoryHello.Hostname` won't be sent to PreHog as it can contain PII but also because it doesn't seem useful):
+
+```protobuf
+message AgentMetadataEvent {
+  string version = 1;
+  string server_id = 2;
+  repeated TeleportAccessProtocol protocols = 3;
+  string os = 4;
+  string os_version = 5;
+  string host_architecture = 6;
+  string glibc_version = 7;
+  repeated string install_methods = 8;
+  string container_runtime = 9;
+  string container_orchestrator = 10;
+  string cloud_environment = 11;
+}
+
+enum TeleportAccessProtocol {
+  TELEPORT_ACCESS_PROTOCOL_UNSPECIFIED = 0;
+  TELEPORT_ACCESS_PROTOCOL_SSH = 1;
+  TELEPORT_ACCESS_PROTOCOL_KUBE = 2;
+  TELEPORT_ACCESS_PROTOCOL_APP = 3;
+  TELEPORT_ACCESS_PROTOCOL_DB = 4;
+  TELEPORT_ACCESS_PROTOCOL_WINDOWS_DESKTOP = 5;
+}
+```
+
+#### Data computation
+
+Both the Teleport version and active Teleport services (which can be used to determine the Teleport access protocols enabled) are already tracked in the inventory control system.
+We detail below how the remaining data will be computed.
+
+##### 3. OS
+
+The OS will be the value on the `GOOS` environment variable.
+This will give us either `linux` or `darwin`.
+
+##### 4. OS version
+
+- `linux`: `$(lsb_release -is) $(lsb_release -rs)` (e.g. "Ubuntu 22.04")
```

Comment:
I'd maybe suggest looking at `/etc/os-release` directly, it's what `lsb_release` reads from and sometimes that script isn't actually present on some systems/images, e.g. the stock `ubuntu:22.04` docker image.

Reply:
Thanks for catching this. The RFD now suggests that we use `/etc/os-release` instead: 4e99b6bf7b23c9077a10fb797ae16756e66e63a9